src/SpinWaveTheory/SpinWaveTheory.jl

```diff
@@ -75,78 +77,79 @@ function generate_local_sun_gens(sys :: System)
-        s_mat_2 = spin_matrices(2)
-        
-        s̃_mat = Array{ComplexF64, 4}(undef, 2, 2, 3, Nₘ)
-
-        no_single_ion = isempty(sys.interactions_union[1].aniso.matrep)
-        T̃_mat = no_single_ion ? zeros(ComplexF64, 0, 0, 0) : Array{ComplexF64, 3}(undef, 2, 2, Nₘ)
-        Q̃_mat = Array{ComplexF64, 4}(undef, 2, 2, 5, Nₘ)
-
-        U_mat_2 = Matrix{ComplexF64}(undef, 2, 2)
-        U_mat_N = Matrix{ComplexF64}(undef, N, N)
-
-        for atom = 1:Nₘ
-            θ, ϕ = dipole_to_angles(sys.dipoles[1, 1, 1, atom])
-            U_mat_N[:] = exp(-1im * ϕ * s_mat_N[3]) * exp(-1im * θ * s_mat_N[2])
-            U_mat_2[:] = exp(-1im * ϕ * s_mat_2[3]) * exp(-1im * θ * s_mat_2[2])
-            @assert isapprox(U_mat_N * U_mat_N', I) "rotation matrix from (global frame to local frame) not unitary"
-            @assert isapprox(U_mat_2 * U_mat_2', I) "rotation matrix from (global frame to local frame) not unitary"
-            for μ = 1:3
-                s̃_mat[:, :, μ, atom] = Hermitian(U_mat_2' * s_mat_2[μ] * U_mat_2)
-            end
-            for ν = 1:5
-                Q̃_mat[:, :, ν, atom] = Hermitian(U_mat_N' * Q_mat_N[ν] * U_mat_N)[1:2, 1:2]
-            end
-
-            if !no_single_ion
-                T̃_mat[:, :, atom] = Hermitian(U_mat_N' * sys.interactions_union[atom].aniso.matrep * U_mat_N)[1:2, 1:2]
-            end
+        for ν = 1:5
+            Q̃_mat[:, :, ν, atom] = Hermitian(U_mat' * Q_mat_N[ν] * U_mat)
         end
+        T̃_mat[:, :, atom] = Hermitian(U_mat' * sys.interactions_union[atom].aniso.matrep * U_mat)
     end
 
     return s̃_mat, T̃_mat, Q̃_mat
 end
 
-function SpinWaveTheory(sys::System{N}, energy_ϵ::Float64=1e-8, energy_tol::Float64=1e-6) where N
-    (N == 0) && error("`SpinWaveTheory` requires an SU(N) `System`. Dipole mode is not currently supported.") 
-
+function SpinWaveTheory(sys::System{N}; energy_ϵ::Float64=1e-8, energy_tol::Float64=1e-6) where N
     # Reshape into single unit cell
     cellsize_mag = cell_dimensions(sys) * diagm(collect(sys.latsize))
     sys = reshape_geometry_aux(sys, (1,1,1), cellsize_mag)
 
-    s̃_mat, T̃_mat, Q̃_mat = generate_local_sun_gens(sys)
+    # Computes the Stevens operator in the local reference frame and the SO(3) rotation matrix from global to local frame
+    # (:dipole mode only)
+    c′_coef = Vector{StevensExpansion}()
+    R_mat   = Vector{Mat3}()
+    if sys.mode == :SUN
+        s̃_mat, T̃_mat, Q̃_mat = generate_local_sun_gens(sys)
+    elseif sys.mode == :dipole
+        Nₘ, Ns = length(sys.dipoles), sys.Ns[1] # number of magnetic atoms and dimension of Hilbert 
+        s_mat_N = spin_matrices(Ns)
+        Nₘ = length(sys.dipoles)
+        s̃_mat = Array{ComplexF64, 4}(undef, 2, 2, 3, Nₘ)
+        T̃_mat = zeros(ComplexF64, 0, 0, 0)
+        Q̃_mat = zeros(ComplexF64, 0, 0, 0, 0)
+        Nₘ = length(sys.dipoles)
+        R  = zeros(Float64, 3, 3)
+        for atom = 1:Nₘ
+            θ, ϕ = dipole_to_angles(sys.dipoles[1, 1, 1, atom])
```

Comment:
This function, `dipole_to_angles`, presumably because of singularities in the inverse trig functions. Does it ever fail? Can we put in conditions instead that handle the bad situations?

Reply:
I had this same concern. Upon closer investigation, it looks like the azimuthal angle `ϕ` is defined with `atan(x, y)`, and this promises to always return a number in $[-\pi, \pi]$. I.e., it makes some particular choice in cases where the azimuthal angle is ambiguous. Maybe we should still verify that the LSWT maintains the correct behavior in this ambiguous case (is it spins aligned with z?)